### PR TITLE
Add persistent cache for toolchain auto-detection mechanism

### PR DESCRIPTION
**Pull request by vmadalin**
### Context
The `Daemon toolchain auto-detection mechanism` is part of `Phase 1` of `toolchain support for Gradle Daemon`, where the motivation behind it and other technical details can be found on the public [spec document](https://docs.google.com/document/d/1CU1e4QjPs2yj_SmJuykPUXutynvn4X_SF0BPXVnAlKI/edit?usp=drive_link). This PR implements the [persistent cache](https://github.com/vmadalin/gradle/issues/4) for the toolchain auto-detection mechanism by adding indexed cache on [CachingJvmMetadataDetector](https://github.com/gradle/gradle/blob/master/platforms/jvm/jvm-services/src/main/java/org/gradle/internal/jvm/inspection/CachingJvmMetadataDetector.java) storing the [JvmInstallationMetadata](https://github.com/gradle/gradle/blob/master/platforms/jvm/jvm-services/src/main/java/org/gradle/internal/jvm/inspection/JvmInstallationMetadata.java) resolved object based on toolchain installation path. This cache is a [VersionStrategy.CachePerVersion](https://github.com/gradle/gradle/blob/67ed0aee543e1043bacadfc4502d2957715a1903/platforms/core-execution/persistent-cache/src/main/java/org/gradle/cache/internal/VersionStrategy.java#L23C17-L23C17) allowing to have a separate cache instance for each Gradle version that are stored under [Gradle User Home directory](https://docs.gradle.org/current/userguide/directory_layout.html#dir:gradle_user_home).

#### Testing
Unit tests:
- [CachingJvmMetadataDetectorTest.groovy](https://github.com/vmadalin/gradle/compare/feature/daemon-toolchain-auto-detection-cli...vmadalin:gradle:feature/daemon-toolchain-auto-detection-cache?expand=1#diff-cde60e3a241fc0816addd8cab97d31130586fc7f9272c7752eb417bf9f91f642)
- [InvalidJvmInstallationCacheInvalidatorTest.groovy](https://github.com/vmadalin/gradle/pull/46/files#diff-2619e8bc0fd354b79c1710f6d99a7a0dabe034f8fdcbf02ad1f89383cb5c771c)

Integration tests
- [CachingJvmMetadataDetectorIntegrationTest.groovy](https://github.com/vmadalin/gradle/compare/feature/daemon-toolchain-auto-detection-cli...vmadalin:gradle:feature/daemon-toolchain-auto-detection-cache?expand=1#diff-3cf52a199a4d544ca11f618687021accbb8cb2e91e69808d68d577d515196eeb)
- [InvalidJvmInstallationReportingIntegrationTest.groovy](https://github.com/vmadalin/gradle/pull/46/files#diff-117b093d72067f5d85cfddffe445dfd18ed8bf48c8f09eea092e27ea37bb203e)
- [DaemonToolchainIntegrationTest.groovy](https://github.com/vmadalin/gradle/pull/46/files#diff-5fa2369a230dc8cf0eccd8bed69a095a9150ad2dcf6fc54379e8ec8c0440924c)

#### Demo
Simple demo exposing how persistent cache for toolchain auto-detection mechanism works:

https://github.com/gradle/gradle/assets/18151158/c00dba91-b56e-4d2b-ad8c-ec03aae54bdb

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### CODE REVIEW

**Comment by vmadalin** [comment]
This PR requires to be able to reuse the `GlobalScopedCacheBuilderFactory` on the `JvmInstallationMetadata` as part of `jvm-services` used from the launcher. However, this isn't longer possible due to https://github.com/gradle/gradle/pull/28582 changes that moved the `persistent-cache` to `core`. Meaning that before merging this PR is required to address to make the mentioned class accessible. CC: @lptr , @big-guy

File: platforms/core-execution/persistent-cache/src/main/java/org/gradle/cache/internal/scopes/DefaultCacheScopeMapping.java [hunk 1]
```diff
@@ -32,11 +31,11 @@ public class DefaultCacheScopeMapping implements CacheScopeMapping {
     private static final Pattern CACHE_KEY_NAME_PATTERN = Pattern.compile("\\p{Alpha}+[-/.\\w]*");
 
     private final File globalCacheDir;
-    private final GradleVersion version;
+    private final String gradleVersion;
```
Comment (lptr) [inline platforms/core-execution/persistent-cache/src/main/java/org/gradle/cache/internal/scopes/DefaultCacheScopeMapping.java:34]: :x: This is not a version now, but a generic string key. I'd rather keep this class bound to `GradleVersion` and move it to some other module that's shared in the right context. If we want to move it here, I suggest we call it `NamedCacheScopeMapping` and update it accordingly.

File: platforms/core-runtime/base-services/src/main/java/org/gradle/cache/internal/scopes/NamedCacheScopeMapping.java [hunk 2]
```diff
@@ -19,24 +19,23 @@
...
 
     private final File globalCacheDir;
-    private final GradleVersion version;
+    private final String gradleVersion;
 
-    public DefaultCacheScopeMapping(File rootDir, GradleVersion version) {
+    public NamedCacheScopeMapping(File rootDir, String gradleVersion) {
         this.globalCacheDir = rootDir;
-        this.version = version;
+        this.gradleVersion = gradleVersion;
```
Comment (lptr) [inline platforms/core-runtime/base-services/src/main/java/org/gradle/cache/internal/scopes/NamedCacheScopeMapping.java:38]: :x: Let's rename this to `name` or something.

**Comment by big-guy** [comment]
Superseded by https://github.com/gradle/gradle/pull/30433

**Pull request closed without merging**